lib/Sema/MiscDiagnostics.cpp

```diff
@@ -245,6 +250,42 @@ static void diagSyntacticUseRestrictions(TypeChecker &TC, const Expr *E,
         }
       }
 
+      // Diagnose tuple expressions with duplicate element label
+      if (auto *tupleExpr = dyn_cast<TupleExpr>(E)) {
+        // FIXME: Duplicate labels on enum payloads should be diagnosed
+        // when declared, not when called.
+        bool isEnumCase = false;
+        if (funcCallExpr) {
+          auto calledValue = funcCallExpr->getCalledValue();
+          if (calledValue) {
+            isEnumCase = isa<EnumElementDecl>(calledValue);
+          }
+        }
+
+        if (!funcCallExpr || isEnumCase) {
```

Comment:
What if this is a subscript index? I believe this will incorrectly diagnose `s[foo: 5, foo: 5]`. In addition, because `funcCallExpr` will be set when visiting other sub-expressions, I believe this won't catch things like:

```swift
func foo(_ fn: () -> Void) {}
foo { print((bar: "", bar: 5).bar) }
```

Ideally the check you want here is `if (!CallArgs.count(tupleExpr))` (which would let you merge this logic with the previous `if` block checking for single-element tuples). Although that will exclude enum case constructor calls, so you'd either want to check if the `Parent` is a `CallExpr` for an enum element decl, or ideally diagnose them separately on the decl.

Reply:
Oh yeah, I'll fix that once the source compatibility tests pass.

Reply:
Fixed 👍